source/common/http/alternate_protocols_cache_impl.h

```diff
@@ -48,12 +49,22 @@ class AlternateProtocolsCacheImpl : public AlternateProtocolsCache {
   // Time source used to check expiration of entries.
   TimeSource& time_source_;
 
-  // Map from hostname to list of alternate protocols.
-  // TODO(RyanTheOptimist): Add a limit to the size of this map and evict based on usage.
-  std::map<Origin, std::vector<AlternateProtocol>> protocols_;
+  struct OriginProtocols {
+    OriginProtocols(const Origin& origin, const std::vector<AlternateProtocol>& protocols)
+        : origin_(origin), protocols_(protocols) {}
+
+    Origin origin_;
+    std::vector<AlternateProtocol> protocols_;
+  };
+  // List of origin, alternate protocol pairs, in insertion order.
+  std::list<OriginProtocols> protocols_list_;
+  // Map from hostname to iterator into protocols_list_.
+  std::map<Origin, std::list<OriginProtocols>::iterator> protocols_map_;
```

Comment:
since we already depend on quiche for the spdy wire parsing, what do you think of also pulling in quiche's linked hash map?

Reply:
Oh, good call! I didn't even consider that. It's a much simpler change. Thanks! Done.